### ISSUE TITLE
Different behavior of the `let_underscore_lock` lint in rustc compared to clippy

### ISSUE THREAD

**Issue by kamulos** [rust-lang/rust#102771]
I tried this code with the new rustc lint:

```rust
fn main() {
    let _ = std::sync::Mutex::new(()).lock();
}
```

Clippy warns here, but rustc doesn't. This might be intentional, but I think this could be a source of mistakes...

The rustc lint is triggered, as soon as I use `.lock().unwrap()`

### Meta

`rustc --version --verbose`:
```
rustc 1.66.0-nightly (0ca356586 2022-10-06)
binary: rustc
commit-hash: 0ca356586fed56002b10920fd21ddf6fb12de797
commit-date: 2022-10-06
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.2
```

**Comment by kadiwa4** [comment]
Fixed by #119710:
```rust
fn main() {
    let _ = parking_lot::Mutex::new(()).lock();
    let _ = std::sync::Mutex::new(()).lock();
}
```
results in
<details><summary>result</summary>

```
error: non-binding `let` on a synchronization lock
 --> src/main.rs:2:5
  |
2 |     let _ = parking_lot::Mutex::new(()).lock();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: consider using an underscore-prefixed named binding or dropping explicitly with `std::mem::drop`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_underscore_lock
  = note: `#[deny(clippy::let_underscore_lock)]` on by default

error: non-binding let on a synchronization lock
 --> src/main.rs:3:9
  |
3 |     let _ = std::sync::Mutex::new(()).lock();
  |         ^ this lock is not assigned to a binding and is immediately dropped
  |
  = note: `#[deny(let_underscore_lock)]` on by default
help: consider binding to an unused variable to avoid immediately dropping the value
  |
3 |     let _unused = std::sync::Mutex::new(()).lock();
  |         ~~~~~~~
help: consider immediately dropping the value
  |
3 |     drop(std::sync::Mutex::new(()).lock());
  |     ~~~~~                                +

error: could not compile `playground` (bin "playground") due to 2 previous errors
```
</details>

Tested in [tests/ui/lint/let_underscore/let_underscore_lock.rs](https://github.com/rust-lang/rust/blob/6db96de66c2c0ea3f4f2f348ed1a83c2c507687d/tests/ui/lint/let_underscore/let_underscore_lock.rs#L12)

I can't remove `needs-triage-legacy` unfortunately.

**Comment by fmease** [comment]
Closing as fixed.
